presto-main/src/main/java/io/prestosql/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -310,6 +311,7 @@ private static void updateQueryOutputLocations(QueryStateMachine queryStateMachi
 
         Optional<int[]> bucketToPartition;
         PartitioningHandle partitioningHandle = plan.getFragment().getPartitioning();
+        AtomicReference<Supplier<Collection<TaskStatus>>> childTaskStatusSupplier = new AtomicReference<>();
```

Comment:
We could avoid `AtomicReference` at all if child stages were created before `SOURCE_DISTRIBUTION` parent stage. This can be done because only dependency seems to be `bucketToPartition` which gets initialized to `Optional.of(new int[1]);` in case of `SOURCE_DISTRIBUTION`.

Perhaps you could extract `createChildStages(...)` method.